### PR TITLE
Add support for passing in discovery url for custom environments

### PR DESCRIPTION
**Pull request by jayanth2830**
- A custom discovery URL is required for environments that are not returned in the default global settings discovery url. 
- Allow users to pass in a Discovery URL to the login cmdlet along with the name of a custom environment.

### CODE REVIEW

**Comment by jayanth2830** [comment]
Will add tests in next iteration

File: src/Common/Commands.Common/PowerBISettings.cs [hunk 1]
```diff
@@ -22,18 +22,22 @@ namespace Microsoft.PowerBI.Commands.Common
     public class PowerBISettings : IPowerBISettings
     {
         public const string FileName = "settings.json";
+        public const string DefaultClientId = "ea0616ba-638b-4df5-95b9-636659ae5121";
+        public const string DefaultRedirect = "urn:ietf:wg:oauth:2.0:oob";
```
Comment (CattieCat) [inline src/Common/Commands.Common/PowerBISettings.cs:26]: Where are the values from?

File: src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs [hunk 2]
```diff
@@ -61,7 +67,29 @@ public class ConnectPowerBIServiceAccount : PowerBICmdlet
 
         public override void ExecuteCmdlet()
         {
-            var environment = this.Settings.Environments[this.Environment];
+            IPowerBIEnvironment environment = null;
+
+            // re-initialize settings with custom environments from discovery url if it exists,
+            // otherwise use existing settings.
+            if (!string.IsNullOrEmpty(this.DiscoveryUrl))
+            {
```
Comment (CodeCyclone) [inline src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs:75]: Will you require customers to also specificy -Environment Custom or if they did specify that, enforce they specify -CustomEnvironment and -DiscoveryUrl

Comment (jayanth2830) [reply src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs:75]: Specifying -Environment Custom is not required. If discovery url was supplied, then CustomEnvironment is a required parameter and the presence of -Environment is not required and will be ignored.

File: src/Common/Commands.Common/PowerBICmdlet.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public abstract class PowerBICmdlet : PSCmdlet, IPowerBICmdlet
         protected IPowerBILoggerFactory LoggerFactory { get; }
         protected IDataStorage Storage { get; }
         protected IAuthenticationFactory Authenticator { get; }
-        protected IPowerBISettings Settings { get; }
+        protected IPowerBISettings Settings { get; set;  }
```
[the remark this answers is not in the document]
Comment (jayanth2830) [reply src/Common/Commands.Common/PowerBICmdlet.cs]: Moved logic to the cmdlet itself so settings can be immutable.

File: src/Common/Commands.Common/PowerBISettings.cs [hunk 4]
```diff
@@ -22,18 +22,22 @@ namespace Microsoft.PowerBI.Commands.Common
     public class PowerBISettings : IPowerBISettings
     {
         public const string FileName = "settings.json";
+        public const string DefaultClientId = "ea0616ba-638b-4df5-95b9-636659ae5121";
```
[the remark this answers is not in the document]
Comment (jayanth2830) [reply src/Common/Commands.Common/PowerBISettings.cs]: moved to settings.json

File: src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs [hunk 5]
```diff
@@ -61,7 +67,29 @@ public class ConnectPowerBIServiceAccount : PowerBICmdlet
...
+                if (string.IsNullOrEmpty(this.CustomEnvironment))
+                {
+                    throw new Exception("CustomEnvironment is required when using a discovery url");
+                }
+
+                this.Settings = new PowerBISettings(null, this.DiscoveryUrl);
+
+                if (!this.Settings.CustomEnvironments.ContainsKey(this.CustomEnvironment))
+                {
+                    throw new Exception($"Discovery URL {this.DiscoveryUrl} did not return environment {this.CustomEnvironment}");
```
[the remark this answers is not in the document]
Comment (jayanth2830) [reply src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs]: updated with latest commit

File: src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs [hunk 6]
```diff
@@ -61,7 +67,29 @@ public class ConnectPowerBIServiceAccount : PowerBICmdlet
...
-            var environment = this.Settings.Environments[this.Environment];
+            IPowerBIEnvironment environment = null;
+
+            // re-initialize settings with custom environments from discovery url if it exists,
+            // otherwise use existing settings.
+            if (!string.IsNullOrEmpty(this.DiscoveryUrl))
+            {
+                if (string.IsNullOrEmpty(this.CustomEnvironment))
+                {
+                    throw new Exception("CustomEnvironment is required when using a discovery url");
```
[the remark this answers is not in the document]
Comment (jayanth2830) [reply src/Modules/Profile/Commands.Profile/ConnectPowerBIServiceAccount.cs]: updated

**Pull request merged**